usecases/modulecomponents/batch/batch.go

```diff
@@ -124,108 +159,210 @@ func (b *Batch) batchWorker() {
 			objCounter++
 		}
 
-		for objCounter < len(job.texts) {
-			if job.ctx.Err() != nil {
-				for j := objCounter; j < len(job.texts); j++ {
-					if !job.skipObject[j] {
-						job.errs[j] = fmt.Errorf("context deadline exceeded or cancelled")
-					}
-				}
+		// if we have a high rate limit we can send multiple batches in parallel.
+		//
+		// If the rate limit is high enough to "fit" the current batch, we send it concurrently. If not, we wait for
+		// either
+		//   - the rate limit to refresh, so we can schedule another concurrent batch
+		//   - the current batch to finish, so the next batch can be sent sequentially
+		//
+		// While using the same code both modes are working slightly different:
+		// 	- For concurrent batching, the amount of used tokens/requests is reserved as long as the batch is running
+		//	  and is cleared when it finishes. This ensures that we never exceed the rate limit and don't need to check
+		//	  the rate limit in the sendBatch function (we use a dummy that never fails a check). All updates happen
+		//	  in the main loop.
+		//  - For sequential batching, the rate limit  will be passed into the sendBatch function and is observed and
+		//    updated there. This allows to use the rate-limit in an optimal way, but also requires more checks. No
+		//    concurrent batch can be started while a sequential batch is running.
+
+		for {
+			timePerToken, objectsPerBatch = b.updateState(rateLimitPerApiKey, timePerToken, objectsPerBatch)
+			numRequests := 1 + int(1.25*float32(len(job.texts)))/objectsPerBatch // round up to be on the safe side
+			if rateLimit.CanSendFullBatch(numRequests, job.tokenSum) {
+				jobCopy := job.copy()
+				rateLimit.ReservedRequests += numRequests
+				rateLimit.ReservedTokens += job.tokenSum
+				timePerToken := timePerToken
+				numRequests := numRequests
+				enterrors.GoWrapper(func() {
+					b.sendBatch(jobCopy, objCounter, dummyRateLimit(), timePerToken, numRequests, true)
+				}, b.logger)
+				break
+			} else if b.concurrentBatches.Load() < 1 {
+				// block so no concurrent batch can be sent
+				b.sendBatch(job, objCounter, rateLimit, timePerToken, 0, false)
 				break
 			}
+			time.Sleep(100 * time.Millisecond)
+		}
+		b.concurrentBatches.Add(1)
```

Comment:
can this be moved to `if rateLimit.CanSendFullBatch`? and decremented in goroutine. Would be easier to reason what this counter is used for. Right now it seems to be decremented first, then incremented.

Reply:
> can this be moved to if rateLimit.CanSendFullBatch?

Sadly not.

The for {}-loop will run until a batch is submited either concurrently (first if) or sequentially (if else) so `rateLimit.CanSendFullBatch` can run repeatedly. I will move the increment to both branches to make it clearer